src/backend/distributed/planner/recursive_planning.c

```diff
@@ -599,6 +616,303 @@ RecursivelyPlanNonColocatedSubqueriesInWhere(Query *query,
 }
 
 
+/*
+ * RecursivelyPlanRecurringTupleOuterJoins descends into a join tree and
+ * recursively plans all non-recurring (i.e., distributed) rels that that
+ * participate in an outer join expression together with a recurring rel,
+ * such as <distributed> in "<recurring> LEFT JOIN <distributed>", i.e.,
+ * where the recurring rel causes returning recurring tuples from the worker
+ * nodes.
+ */
+static void
+RecursivelyPlanRecurringTupleOuterJoins(Node *node, Query *query,
+										RecursivePlanningContext *recursivePlanningContext)
+{
+	if (node == NULL)
+	{
+		return;
+	}
+	else if (IsA(node, RangeTblRef))
+	{
+		return;
+	}
+	else if (IsA(node, FromExpr))
+	{
+		FromExpr *fromExpr = (FromExpr *) node;
+		ListCell *fromExprCell;
+
+		/* search for join trees in each FROM element */
+		foreach(fromExprCell, fromExpr->fromlist)
+		{
+			Node *fromElement = (Node *) lfirst(fromExprCell);
+
+			RecursivelyPlanRecurringTupleOuterJoins(fromElement, query,
+													recursivePlanningContext);
+		}
+
+		return;
+	}
+	else if (IsA(node, JoinExpr))
+	{
+		JoinExpr *joinExpr = (JoinExpr *) node;
+
+		Node *leftNode = joinExpr->larg;
+		Node *rightNode = joinExpr->rarg;
+
+		/* there may be recursively plannable outer joins deeper in the join tree */
+		RecursivelyPlanRecurringTupleOuterJoins(joinExpr->larg, query,
+												recursivePlanningContext);
+		RecursivelyPlanRecurringTupleOuterJoins(joinExpr->rarg, query,
+												recursivePlanningContext);
+		switch (joinExpr->jointype)
+		{
+			case JOIN_LEFT:
+			{
+				bool leftNodeRecurs = IsJoinNodeRecurring(leftNode, query);
+				bool rightNodeRecurs = IsJoinNodeRecurring(rightNode, query);
+
+				/* <recurring> left join <distributed> */
+				if (leftNodeRecurs && !rightNodeRecurs)
+				{
+					RecursivelyPlanDistributedJoinNode(rightNode, query,
```

Comment:
I guess DEBUG1/2 message would be nice when we decide to recursively plan for outer joins?